packages/contracts/contracts/optimistic-ethereum/ovm/StateManagerGasSanitizer.sol

```diff
@@ -0,0 +1,328 @@
+    uint constant SET_STORAGE_VIRTUAL_GAS_COST = 20000;
+    // Nonces
+    uint constant GET_CONTRACT_NONCE_VIRTUAL_GAS_COST = 10000;
+    uint constant SET_CONTRACT_NONCE_VIRTUAL_GAS_COST = 30000;
+    uint constant INCREMENT_CONTRACT_NONCE_VIRTUAL_GAS_COST = 35000;
+    // Code
+    uint constant ASSOCIATE_CODE_CONTRACT_VIRTUAL_GAS_COST = 1000;
+    uint constant REGISTER_CREATED_CONTRACT_VIRTUAL_GAS_COST = 1000;
+    uint constant GET_CODE_CONTRACT_ADDRESS_VIRTUAL_GAS_COST = 1000;
+    uint constant GET_CODE_CONTRACT_HASH_VIRTUAL_GAS_COST = 1000;
+    // Code copy retrieval, linear in code size
+    uint constant GET_CODE_CONTRACT_BYTECODE_VIRTUAL_GAS_COST_PER_BYTE = 10;
+
+    /*
+     * Constant/Upper-bounded Fixed Gas Cost Constants
+     */
+
+    // Storage
+    uint constant GET_STORAGE_GAS_COST_UPPER_BOUND = 200000;
+    uint constant SET_STORAGE_GAS_COST_UPPER_BOUND = 200000;
+    // Nonces
+    uint constant GET_CONTRACT_NONCE_GAS_COST_UPPER_BOUND = 200000;
+    uint constant SET_CONTRACT_NONCE_GAS_COST_UPPER_BOUND = 200000;
+    uint constant INCREMENT_CONTRACT_NONCE_GAS_COST_UPPER_BOUND = 200000;
+    // Code
+    uint constant ASSOCIATE_CODE_CONTRACT_GAS_COST_UPPER_BOUND = 200000;
+    uint constant REGISTER_CREATED_CONTRACT_GAS_COST_UPPER_BOUND = 200000;
+    uint constant GET_CODE_CONTRACT_ADDRESS_GAS_COST_UPPER_BOUND = 200000;
+    uint constant GET_CODE_CONTRACT_HASH_GAS_COST_UPPER_BOUND = 200000;
+    // Code copy retrieval, linear in code size
+    uint constant GET_CODE_CONTRACT_BYTECODE_GAS_COST_UPPER_BOUND = 200000;
+
+
+    /*
+     * Contract Variables
+     */
+
+    uint OVMRefund;
+
+    /*
+     * Modifiers
+     */
+
+
+    /*
+     * Constructor
+     */
+
+    /**
+     * @param _addressResolver Address of the AddressResolver contract.
+     */
+    constructor(
+        address _addressResolver
+    )
+        public
+        ContractResolver(_addressResolver)
+    {}
+
+    /*
+     * Gas Virtualization and Storage
+     */
+
+    // External Initialization and Retrieval Logic
+    function resetOVMRefund() external {
+        OVMRefund = 0;
+    }
+
+    function getOVMRefund() external view returns(uint) {
+        return OVMRefund;
+    }
+
+    // Internal Logic
+    function addToOVMRefund(uint _refund) internal {
+        OVMRefund += _refund;
+        return;
+    }
+
+    /** TODO UPDATE THIS DOCSTR
+     * Forwards a call to this proxy along to the actual state manager, and consumes any leftover gas up to _virtualGasCost.
+     */
+    function performSanitizedProxyAndRecordRefund(
+        uint _sanitizedGasCost,
+        uint _virtualGasCost
+    ) internal {
+        uint initialGas = gasleft();
+        address stateManager = resolveStateManager();
+
+        uint refund = _sanitizedGasCost - _virtualGasCost;
+        addToOVMRefund(refund);
+
+        bool success;
+        uint returnedSize;
+        uint returnDataStart;
+        assembly {
+            let initialFreeMemStart := mload(0x40)
```

Comment:
Might suggest naming this what it represents like `callBytes` or `callBytesPointer`-- ideally all newly allocated  memory is valid and starts with free mem.